### DISCUSSION TITLE
Failed challenge transactions

### DISCUSSION THREAD

**Discussion by 0xSimeon** [Cyfrin/foundry-full-course-cu#59] [Q&A] [answered]
I'm trying to solve the challenge for lesson 4 and anytime I call the `solveChallenge()`, metamask can't estimate gas and the transaction ultimately fails. 

Reference tx: https://sepolia.etherscan.io/tx/0x2885311039525be370b8c0caebf146c1566d9267d0b54e4c92655ea9f30861bf

**Comment by cromewar** [comment]
Hello @0xSimeon if you get the estimate gas error, it means you are not solving the problem properly, don't give up. We can give you hints if you explain what you've done.

**Reply by 0xSimeon** [reply]
Thanks for your response. To solve the challenge, I called `solveChallenge()` with two arguments.
- My guess `priceGuess` 
- My twitter username.

**Reply by palmcivet7** [reply]
I'm having a similar issue. This is what I'm doing:
- call getPriceFeed on Lesson 4 contract
- call latestRoundData on i_priceFeed
- use answer int256 for my priceGuess

This is causing a gas error and I have no idea what I'm doing wrong.

**Reply by cromewar** [reply]
@alymurtazamemon Have you any clue here?

**Reply by PatrickAlphaC** [reply]
Just remember guys - the testnet price feed doesn't always match the live price feeds!!! Where could you read the actual price in? 🤔

**Reply by alymurtazamemon** [reply]
> @alymurtazamemon Have you any clue here?

Yes

@0xSimeon @palmcivet7 Make sure two things, 1. you are reading the correct price feed according to the network, 2. These price feeds change the price data after the heartbeat rate you need to be quick here to enter the price before the new price update.

HINT: Give focus on this line 
```solidity
if (actualPrice + 3e18 > priceGuess && actualPrice - 3e18 < priceGuess) {
```

**Reply by heysourin** [reply]
@alymurtazamemon did you clear this challenge?

**Reply by alymurtazamemon** [reply]
> @alymurtazamemon did you clear this challenge?

yes

**Reply by heysourin** [reply]
@alymurtazamemon I couldnot solve it, tried a lot. please send me the answer and explanation. pls dm on twitter @heysourin or email sourin07@gmail.com

**Reply by palmcivet7** [reply]
> @0xSimeon @palmcivet7 Make sure two things, 1. you are reading the correct price feed according to the network, 2. These price feeds change the price data after the heartbeat rate you need to be quick here to enter the price before the new price update.
> 
> HINT: Give focus on this line
> 
> ```solidity
> if (actualPrice + 3e18 > priceGuess && actualPrice - 3e18 < priceGuess) {
> ```

Thanks for the response, but I still don't understand. Does it mean I should be changing the answer from the latestRoundData? Surely not?

I'm getting the priceFeed address from here in the Lesson 4 contract:
```
 function getPriceFeed() external view returns (address) {
        return i_priceFeed;
    }
```
Then I'm getting the answer from here in the priceFeed contract:
```
  function latestRoundData()
    external
    view
    returns (
      uint80 roundId,
      int256 answer,
      uint256 startedAt,
      uint256 updatedAt,
      uint80 answeredInRound
    );
```
This answer should be the same as the uint that is returned from here in the Lesson 4 contract:
```
 function getPrice() private view returns (uint256) {
        (, int256 price,,,) = AggregatorV3Interface(i_priceFeed).latestRoundData();
        return uint256(price);
    }
```
The solve challenge function declares:
```
uint256 actualPrice = getPrice();
```
So my understanding is actualPrice should be the same number I'm getting from the latestRoundData in the priceFeed contract, and with this line:
```
 if (actualPrice + 3e18 > priceGuess && actualPrice - 3e18 < priceGuess)
```
The answer I'm giving can be slightly less or more than the actualPrice number, which is, as far as I can tell, coming from the exact same source as the answer I'm giving. What am I missing? I have two tabs open between read priceFeed contract and write Lesson 4 contract so I'm confident it's not a delay issue. This is all on Arbiscan by the way.

**Reply by alymurtazamemon** [reply]
@0xSimeon Can you show the price feed address?

**Reply by palmcivet7** [reply]
https://arbiscan.io/address/0x639fe6ab55c921f74e7fac1ee960c0b6293ba612#readContract

**Reply by heysourin** [reply]
@alymurtazamemon Bro if you have solution, please send!!!

**Reply by alymurtazamemon** [reply]
> https://arbiscan.io/address/0x639fe6ab55c921f74e7fac1ee960c0b6293ba612#readContract

Address it correct.

**Reply by heysourin** [reply]
@alymurtazamemon Bro did you really solve it?

**Reply by palmcivet7** [reply]
Thanks, that's good to know. Can you see where I'm going wrong?

**Reply by alymurtazamemon** [reply]
> @alymurtazamemon Bro did you really solve it?

😅 

https://opensea.io/assets/arbitrum/0x39338138414df90ec67dc2ee046ab78bcd4f56d9/30

**Reply by heysourin** [reply]
> > @alymurtazamemon Bro did you really solve it?
> 
> 😅
> 
> https://opensea.io/assets/arbitrum/0x39338138414df90ec67dc2ee046ab78bcd4f56d9/30

nice 📎, then pls send the solution and explanation bro

**Reply by alymurtazamemon** [reply]
> Thanks, that's good to know. Can you see where I'm going wrong?

In the last step.

HINT: Your this line is not true;
```
The answer I'm giving can be slightly less or more than the actualPrice number, which is, as far as I can tell, coming from the exact same source as the answer I'm giving.
```

Please do not ask more questions because I will not answer this is your challenge.

**Reply by palmcivet7** [reply, marked as answer]
>Please do not ask more questions because I will not answer this is your challenge.

That's ok, thanks for your help so far. I will continue to use this thread to think out loud.

>this line is not true

Let's break down the line:
>```
>The answer I'm giving can be slightly less or more than the actualPrice number
>```
I said this because of this line in the code:
```
if (actualPrice + 3e18 > priceGuess && actualPrice - 3e18 < priceGuess)
```
"slightly less or more" being no more than actualPrice + 3e18 and no less than actualPrice - 3e18. The priceGuess needs to fit into this range.
Let's look at the other part of the line:
>```
>actualPrice (is) coming from the exact same source as the answer I'm giving
>```
The actualPrice comes from this line in the code:
```
uint256 actualPrice = getPrice();
```
getPrice() comes from this line in the code:
```
function getPrice() private view returns (uint256) {
        (, int256 price,,,) = AggregatorV3Interface(i_priceFeed).latestRoundData();
        return uint256(price);
    }
```
int256 price comes from latestRoundData in the priceFeed contract (https://arbiscan.io/address/0x639fe6ab55c921f74e7fac1ee960c0b6293ba612#code): 
```
function latestRoundData()
    public
    view
    virtual
    override
    returns (
      uint80 roundId,
      int256 answer,
      uint256 startedAt,
      uint256 updatedAt,
      uint80 answeredInRound
    )
  {
    Phase memory current = currentPhase; // cache storage reads

    (
      uint80 roundId,
      int256 answer,
      uint256 startedAt,
      uint256 updatedAt,
      uint80 ansIn
    ) = current.aggregator.latestRoundData();

    return addPhaseIds(roundId, answer, startedAt, updatedAt, ansIn, current.id);
  }
  ```
To rephrase what I previously said; my guess can be within the range of (actualPrice + 3e18 > priceGuess && actualPrice - 3e18 < priceGuess). The actualPrice is within this range, therefore if my priceGuess is the actualPrice, it should satisfy this condition. And yet it's not because something about this is untrue, but I can't see what.

**Edit:** I finally managed to solve this and you _do_ have to adjust the number you submit.

**Reply by heysourin** [reply]
@palmcivet7 Me too, facing the same issue!!!
